src/OpenTelemetry/Metrics/Exemplar/AlignedHistogramBucketExemplarReservoir.cs

```diff
@@ -32,26 +32,62 @@ public AlignedHistogramBucketExemplarReservoir(int length)
         this.snapshotExemplars = new Exemplar[length + 1];
     }
 
-    public void OfferAtBoundary(int index, double value)
+    public void OfferAtBoundary(int index, double value, ReadOnlySpan<KeyValuePair<string, object>> tags)
     {
-        var exemplar = default(Exemplar);
+        ref var exemplar = ref this.runningExemplars[index];
         exemplar.Timestamp = DateTime.UtcNow;
         exemplar.DoubleValue = value;
         exemplar.TraceId = Activity.Current?.TraceId;
         exemplar.SpanId = Activity.Current?.SpanId;
-        this.runningExemplars[index] = exemplar;
+
+        if (tags == default)
+        {
+            // default tag is used to indicate
+            // the special case where all tags provided at measurement
+            // recording time are stored.
+            // In this case, Exemplars does not have to store any tags.
+            // In other words, FilteredTags will be empty.
+            return;
+        }
+
+        if (exemplar.FilteredTags == null)
+        {
+            exemplar.FilteredTags = new List<KeyValuePair<string, object>>(tags.Length);
+        }
+        else
+        {
+            // Keep the list, but clear contents.
+            exemplar.FilteredTags.Clear();
+        }
+
+        // Though only those tags that are filtered need to be
+        // stored, finding filtered list from the full tag list
+        // is expensive. So all the tags are stored in hot path (this).
+        // During snapshot, the filtered list is calculated.
+        // TODO: Evaluate alternative approaches based on perf.
+        foreach (var tag in tags)
+        {
+            exemplar.FilteredTags.Add(tag);
+        }
     }
 
     public Exemplar[] Collect()
     {
         return this.snapshotExemplars;
     }
 
-    public void SnapShot(bool reset)
+    public void SnapShot(ReadOnlyTagCollection actualTags, bool reset)
     {
         for (int i = 0; i < this.runningExemplars.Length; i++)
         {
             this.snapshotExemplars[i] = this.runningExemplars[i];
+            if (this.runningExemplars[i].FilteredTags != null)
+            {
+                // TODO: Better data structure to avoid this Linq.
+                // This is doing filtered = alltags - storedtags.
+                this.snapshotExemplars[i].FilteredTags = this.runningExemplars[i].FilteredTags.Except(actualTags.KeyAndValues.ToList()).ToList();
```

Comment:
This might be a good fit for lazy evaluation (e.g. the first exporter which triggered the serialization will actually handle evaluation, though the minor downside is that CPU consumption attribution will be harder).